File: src/main/java/ru/org/linux/comment/CommentPrepareService.java
/*
 * Copyright 1998-2022 Linux.org.ru
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

package ru.org.linux.comment;

import com.google.common.base.Strings;
import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableMap;
import com.google.common.collect.ImmutableSet;
import org.springframework.stereotype.Service;
import ru.org.linux.markup.MessageTextService;
import ru.org.linux.site.ApiDeleteInfo;
import ru.org.linux.site.DeleteInfo;
import ru.org.linux.spring.dao.DeleteInfoDao;
import ru.org.linux.spring.dao.MessageText;
import ru.org.linux.spring.dao.MsgbaseDao;
import ru.org.linux.spring.dao.UserAgentDao;
import ru.org.linux.topic.Topic;
import ru.org.linux.topic.TopicPermissionService;
import ru.org.linux.user.*;

import javax.annotation.Nullable;
import java.util.*;
import java.util.stream.Collectors;

@Service
public class CommentPrepareService {
  private final MessageTextService textService;
  private final MsgbaseDao msgbaseDao;
  private final TopicPermissionService topicPermissionService;
  private final UserService userService;
  private final DeleteInfoDao deleteInfoDao;
  private final UserAgentDao userAgentDao;
  private final RemarkDao remarkDao;

  public CommentPrepareService(MessageTextService textService, MsgbaseDao msgbaseDao,
                               TopicPermissionService topicPermissionService, UserService userService,
                               DeleteInfoDao deleteInfoDao, UserAgentDao userAgentDao, RemarkDao remarkDao) {
    this.textService = textService;
    this.msgbaseDao = msgbaseDao;
    this.topicPermissionService = topicPermissionService;
    this.userService = userService;
    this.deleteInfoDao = deleteInfoDao;
    this.userAgentDao = userAgentDao;
    this.remarkDao = remarkDao;
  }

  private PreparedComment prepareComment(
          MessageText messageText,
          User author,
          Optional<String> remark,
          Comment comment,
          Optional<CommentList> comments,
          Profile profile,
          Topic topic,
          Set<Integer> hideSet,
          Set<Integer> samePageComments,
          @Nullable User currentUser) {
    String processedMessage = textService.renderCommentText(messageText, !topicPermissionService.followAuthorLinks(author));

    ReplyInfo replyInfo = null;
    boolean deletable = false;
    boolean editable = false;
    int answerCount;
    String answerLink;
    boolean answerSamepage = false;

    if (comments.isPresent()) {
      if (comment.getReplyTo() != 0) {
        CommentNode replyNode = comments.get().getNode(comment.getReplyTo());

        boolean replyDeleted = replyNode == null;
        if (replyDeleted) {
          // ответ на удаленный комментарий
          replyInfo = new ReplyInfo(comment.getReplyTo(), true);
        } else {
          Comment reply = replyNode.getComment();

          boolean samePage = samePageComments.contains(reply.getId());

          String replyAuthor = userService.getUserCached(reply.getUserid()).getNick();

          replyInfo = new ReplyInfo(
                  reply.getId(),
                  replyAuthor,
                  Strings.emptyToNull(reply.getTitle().trim()),
                  reply.getPostdate(),
                  samePage,
                  false
          );
        }
      }

      CommentNode node = comments.get().getNode(comment.getId());
      List<CommentNode> replysFiltered = node.childs().stream().filter(commentNode ->
              !hideSet.contains(commentNode.getComment().getId())
      ).collect(Collectors.toList());

      answerCount = replysFiltered.size();

      if (answerCount > 1) {
        answerLink = topic.getLink()+"/thread/" + comment.getId()+"#comments";
      } else if (answerCount == 1) {
        answerLink = topic.getLink()+"?cid=" + replysFiltered.get(0).getComment().getId();
        answerSamepage = samePageComments.contains(replysFiltered.get(0).getComment().getId());
      } else {
        answerLink = null;
      }

      deletable = topicPermissionService.isCommentDeletableNow(comment, currentUser, topic, node.hasAnswers());
      editable = topicPermissionService.isCommentEditableNow(comment, currentUser, node.hasAnswers(), topic, messageText.markup());
    } else {
      answerCount = 0;
      answerLink = null;
    }

    Userpic userpic = null;

    if (profile.isShowPhotos()) {
      userpic = userService.getUserpic(author, profile.getAvatarMode(), false);
    }

    ApiUserRef ref = userService.ref(author, currentUser);

    Optional<DeleteInfo> deleteInfo = loadDeleteInfo(comment);
    Optional<ApiDeleteInfo> apiDeleteInfo = deleteInfo.map(i ->
            new ApiDeleteInfo(userService.getUserCached(i.getUserid()).getNick(), i.getReason())
    );

    EditSummary editSummary = loadEditSummary(comment);

    String postIP = null;
    String userAgent = null;

    if (currentUser!=null && currentUser.isModerator()) {
      postIP = comment.getPostIP();
      userAgent = userAgentDao.getUserAgentById(comment.getUserAgentId());
    }

    boolean undeletable = topicPermissionService.isUndeletable(topic, comment, currentUser, deleteInfo);

    return new PreparedComment(comment, ref, processedMessage, replyInfo,
            deletable, editable, remark.orElse(null), userpic, apiDeleteInfo.orElse(null), editSummary,
            postIP, userAgent, comment.getUserAgentId(), undeletable, answerCount, answerLink, answerSamepage);
  }

  private Optional<DeleteInfo> loadDeleteInfo(Comment comment) {
    if (comment.isDeleted()) {
      return deleteInfoDao.getDeleteInfo(comment.getId());
    } else {
      return Optional.empty();
    }
  }

  private EditSummary loadEditSummary(Comment comment) {
    EditSummary editSummary = null;

    if (comment.getEditCount()>0) {
      editSummary = new EditSummary(
              userService.getUserCached(comment.getEditorId()).getNick(),
              comment.getEditDate(),
              comment.getEditCount()
      );
    }

    return editSummary;
  }

  private PreparedRSSComment prepareRSSComment(MessageText messageText, Comment comment) {
    User author = userService.getUserCached(comment.getUserid());

    String processedMessage = textService.renderTextRSS(messageText);

    return new PreparedRSSComment(comment, author, processedMessage);
  }

  public PreparedComment prepareCommentForReplyto(Comment comment, @Nullable User currentUser, Profile profile, Topic topic) {
    MessageText messageText = msgbaseDao.getMessageText(comment.getId());
    User author = userService.getUserCached(comment.getUserid());

    return prepareComment(messageText, author, Optional.empty(), comment, Optional.empty(), profile, topic, ImmutableSet.of(), ImmutableSet.of(), currentUser);
  }

  /**
   * Подготовить комментарий для последующего редактирования
   * в комментарии не используется автор и не важно существует ли ответ и автор ответа
   * @param comment Редактируемый комментарий
   * @param message Тело комментария
   * @return подготовленный коментарий
   */
  public PreparedComment prepareCommentForEdit(Comment comment, MessageText message) {
    User author = userService.getUserCached(comment.getUserid());
    String processedMessage = textService.renderCommentText(message, false);

    ApiUserRef ref = userService.ref(author, null);

    return new PreparedComment(
        comment,
        ref,
        processedMessage,
        null, // reply
        false, // deletable
        false, // editable
        null,   // Remark
        null, // userpic
        null,
        null,
        null,
        null,
            0,
            false, 0, null, false);
  }

  public List<PreparedRSSComment> prepareCommentListRSS(List<Comment> list) {
    List<PreparedRSSComment> commentsPrepared = new ArrayList<>(list.size());
    for (Comment comment : list) {
      MessageText messageText = msgbaseDao.getMessageText(comment.getId());

      commentsPrepared.add(prepareRSSComment(messageText, comment));
    }
    return commentsPrepared;
  }

  private Map<Integer, User> loadUsers(Iterable<Integer> userIds) {
    ImmutableMap.Builder<Integer, User> builder = ImmutableMap.builder();

    for (User user : userService.getUsersCachedJava(ImmutableSet.copyOf(userIds))) {
      builder.put(user.getId(), user);
    }

    return builder.build();
  }

  public List<PreparedComment> prepareCommentList(
          CommentList comments,
          List<Comment> list,
          Topic topic,
          Set<Integer> hideSet,
          @Nullable User currentUser,
          Profile profile) {
    if (list.isEmpty()) {
      return ImmutableList.of();
    }

    Map<Integer, MessageText> texts = msgbaseDao.getMessageText(list.stream().map(Comment::getId).collect(Collectors.toList()));

    Map<Integer, User> users = loadUsers(list.stream().map(Comment::getUserid).collect(Collectors.toList()));

    Map<Integer, Remark> remarks;

    if (currentUser!=null) {
      remarks = remarkDao.getRemarks(currentUser, users.values());
    } else {
      remarks = ImmutableMap.of();
    }

    Set<Integer> samePageComments = list.stream().map(Comment::getId).collect(Collectors.toSet());

    return list.stream().map(comment -> {
      MessageText text = texts.get(comment.getId());

      User author = users.get(comment.getUserid());

      Optional<Remark> remark = Optional.ofNullable(remarks.get(author.getId()));

      return prepareComment(text, author, remark.map(Remark::getText), comment, Optional.of(comments), profile, topic,
              hideSet, samePageComments, currentUser);
    }).collect(Collectors.toList());
  }

  public List<PreparedCommentsListItem> prepareCommentsList(List<CommentsListItem> comments) {
    Map<Integer, User> users = loadUsers(comments.stream().map(CommentsListItem::getAuthorId).collect(Collectors.toList()));
    Map<Integer, MessageText> texts = msgbaseDao.getMessageText(comments.stream().map(CommentsListItem::getCommentId).collect(Collectors.toList()));

    return comments.stream().map(comment -> {
      User author = users.get(comment.authorId());

      String plainText = textService.extractPlainText(texts.get(comment.getCommentId()));
      String textPreview = MessageTextService.trimPlainText(plainText, 250, false);

      return new PreparedCommentsListItem(comment, author, textPreview);
    }).collect(Collectors.toList());
  }
}

File: src/main/scala/ru/org/linux/comment/CommentPrepareService.scala
/*
 * Copyright 1998-2024 Linux.org.ru
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */
package ru.org.linux.comment

import com.google.common.base.Strings
import org.joda.time.{DateTime, Duration}
import org.springframework.stereotype.Service
import ru.org.linux.auth.AnySession
import ru.org.linux.group.{Group, GroupDao}
import ru.org.linux.markup.MessageTextService
import ru.org.linux.reaction.{PreparedReactions, ReactionService}
import ru.org.linux.site.ApiDeleteInfo
import ru.org.linux.spring.dao.{DeleteInfoDao, MessageText, MsgbaseDao, UserAgentDao}
import ru.org.linux.topic.{Topic, TopicPermissionService}
import ru.org.linux.user.*
import ru.org.linux.warning.{Warning, WarningService}

import scala.jdk.CollectionConverters.*
import scala.jdk.OptionConverters.*

@Service
class CommentPrepareService(textService: MessageTextService, msgbaseDao: MsgbaseDao,
                            topicPermissionService: TopicPermissionService, userService: UserService,
                            deleteInfoDao: DeleteInfoDao, userAgentDao: UserAgentDao, remarkDao: RemarkDao,
                            groupDao: GroupDao, reactionPrepareService: ReactionService,
                            warningService: WarningService) {

  private def prepareComment(messageText: MessageText, author: User, remark: Option[String], comment: Comment,
                             comments: Option[CommentList], topic: Topic, hideSet: Set[Int], samePageComments: Set[Int],
                             group: Group, ignoreList: Set[Int], filterShow: Boolean, warnings: Seq[Warning])
                            (implicit session: AnySession) = {
    val processedMessage = textService.renderCommentText(messageText, !topicPermissionService.followAuthorLinks(author))

    val (answerLink, answerSamepage, answerCount, replyInfo, hasAnswers) = if (comments.isDefined) {
      val replyInfo: Option[ReplyInfo] = if (comment.replyTo != 0) {
        comments.get.getNodeOpt(comment.replyTo) match {
          case None => // ответ на удаленный комментарий
            Some(new ReplyInfo(comment.replyTo, true))
          case Some(replyNode) =>
            val reply = replyNode.getComment
            val samePage = samePageComments.contains(reply.id)
            val replyAuthor = userService.getUserCached(reply.userid).getNick
            Some(new ReplyInfo(reply.id, replyAuthor, Strings.emptyToNull(reply.title.trim), reply.postdate, samePage, false))
        }
      } else {
        None
      }

      val node = comments.get.getNode(comment.id)
      val replysFiltered = node.childs.asScala.filter(commentNode => !hideSet.contains(commentNode.getComment.id))

      val answerCount = replysFiltered.size

      if (answerCount > 1) {
        val link = if (!filterShow) {
          s"${topic.getLink}/thread/${comment.id}#comments"
        } else {
          s"${topic.getLink}/thread/${comment.id}?filter=show#comments"
        }

        (Some(link), false, answerCount, replyInfo, true)
      } else if (answerCount == 1) {
        val answerSamepage = samePageComments.contains(replysFiltered.head.getComment.id)

        val link = if (!filterShow) {
          s"${topic.getLink}?cid=${replysFiltered.head.getComment.id}"
        } else {
          s"${topic.getLink}?filter=show&cid=${replysFiltered.head.getComment.id}"
        }

        (Some(link), answerSamepage, 1, replyInfo, true)
      } else {
        (None, false, 0, replyInfo, false)
      }
    } else {
      (None, false, 0, None, false)
    }

    val userpic: Option[Userpic] = if (session.profile.showPhotos) {
      Some(userService.getUserpic(author, session.profile.avatarMode, misteryMan = false))
    } else {
      None
    }

    val deleteInfo = loadDeleteInfo(comment)
    val apiDeleteInfo = deleteInfo.map(i => new ApiDeleteInfo(userService.getUserCached(i.userid).getNick, i.getReason))
    val editSummary = loadEditSummary(comment)

    val (postIP, userAgent) = if (session != null && session.moderator) {
      (Option(comment.postIP), userAgentDao.getUserAgentById(comment.userAgentId).toScala)
    } else {
      (None, None)
    }

    val undeletable = topicPermissionService.isUndeletable(topic, comment, deleteInfo)(session)
    val deletable = topicPermissionService.isCommentDeletableNow(comment, topic, hasAnswers)(session)
    val editable = topicPermissionService.isCommentEditableNow(comment, hasAnswers, topic, messageText.markup)(session)
    val warningsAllowed = topicPermissionService.canPostWarning(topic, Some(comment))(session)

    val authorReadonly = !topicPermissionService.isCommentsAllowedByUser(group, topic, Some(author), ignoreFrozen = true)

    val preparedWarnings = warningService.prepareWarning(warnings)

    PreparedComment(comment = comment, author = author, processedMessage = processedMessage, reply = replyInfo,
      deletable = deletable, editable = editable, remark = remark, userpic = userpic, deleteInfo = apiDeleteInfo,
      editSummary = editSummary, postIP = postIP, userAgent = userAgent, undeletable = undeletable,
      answerCount = answerCount, answerLink = answerLink, answerSamepage = answerSamepage,
      authorReadonly = authorReadonly,
      reactions = reactionPrepareService.prepare(comment.reactions, ignoreList,topic, Some(comment)),
      warningsAllowed = warningsAllowed, warnings = preparedWarnings)
  }

  private def loadDeleteInfo(comment: Comment) = {
    if (comment.deleted) {
      deleteInfoDao.getDeleteInfo(comment.id).toScala
    } else {
      None
    }
  }

  private def loadEditSummary(comment: Comment): Option[EditSummary] = {
    if (comment.editCount > 0) {
      Some(new EditSummary(userService.getUserCached(comment.editorId).getNick, comment.editDate, comment.editCount))
    } else {
      None
    }
  }

  def prepareCommentOnly(comment: Comment, topic: Topic, ignoreList: Set[Int])
                        (implicit currentUser: AnySession): PreparedComment = {
    assert(comment.topicId == topic.id)

    val messageText = msgbaseDao.getMessageText(comment.id)
    val author = userService.getUserCached(comment.userid)
    val group = groupDao.getGroup(topic.groupId)

    prepareComment(messageText = messageText, author = author, remark = None, comment = comment, comments = None,
      topic = topic, hideSet = Set.empty, samePageComments = Set.empty, group = group,
      ignoreList = ignoreList, filterShow = false, warnings = Seq.empty)
  }

  /**
   * Подготовить комментарий для последующего редактирования
   * в комментарии не используется автор и не важно существует ли ответ и автор ответа
   *
   * @param comment Редактируемый комментарий
   * @param message Тело комментария
   * @return подготовленный коментарий
   */
  def prepareCommentForEdit(comment: Comment, message: MessageText): PreparedComment = {
    val author = userService.getUserCached(comment.userid)
    val processedMessage = textService.renderCommentText(message, nofollow = false)

    PreparedComment(comment = comment, author = author, reply = None, editable = false, remark = None,
      userpic = None, deleteInfo = None, editSummary = None, postIP = None, userAgent = None, undeletable = false,
      answerCount = 0, answerLink = None, answerSamepage = false, authorReadonly = false,
      processedMessage = processedMessage, deletable = false, reactions = PreparedReactions.emptyDisabled,
      warningsAllowed = false, warnings = Seq.empty)
  }

  def prepareCommentList(comments: CommentList, list: Seq[Comment], topic: Topic, hideSet: Set[Int],
                         ignoreList: Set[Int], filterShow: Boolean)
                        (implicit currentUser: AnySession): Seq[PreparedComment] = {
    if (list.isEmpty) {
      Seq.empty
    } else {
      val texts = msgbaseDao.getMessageText(list.map(_.id))
      val users = userService.getUsersCachedMap(list.map(_.userid))
      val group = groupDao.getGroup(topic.groupId)

      val allWarnings: Map[Int, Seq[Warning]] = if (!topic.expired && currentUser.moderator) {
        warningService.load(list)
      } else {
        Map.empty
      }

      val remarks = currentUser.opt.map { user =>
        remarkDao.getRemarks(user.user, users.values)
      }.getOrElse(Map.empty[Int, Remark])

      val samePageComments = list.map(_.id).toSet

      list.map { comment =>
        val text = texts(comment.id)
        val author = users(comment.userid)
        val remark = remarks.get(author.getId)
        val warnings = allWarnings.getOrElse(comment.id, Seq.empty)

        prepareComment(text, author, remark.map(_.getText), comment, Option(comments), topic, hideSet,
          samePageComments, group, ignoreList, filterShow, warnings)
      }
    }
  }

  def buildDateJumpSet(comments: Seq[Comment], jumpMinDuration: Duration): java.util.Set[Integer] = {
    val commentDates = comments.view.map { c =>
      c.id -> new DateTime(c.postdate)
    }

    commentDates.zip(commentDates.drop(1)).filter { case (first, second) =>
      new Duration(first._2, second._2).isLongerThan(jumpMinDuration)
    }.map(_._2._1).map(Integer.valueOf).toSet.asJava
  }
}